### PR TITLE
Fix #5533: Prefer light_rewrite_comment if it is not a doccomment

### LINKED ISSUE #5533: `format_code_in_doc_comments` formats some long line comments to block comments
## Meta

- **rustfmt version**: `rustfmt 1.5.1-nightly (b44197ab 2022-09-05)`
- **Minimum reproducible repo**: <https://github.com/pan93412/rustfmt-format-code-in-doc-comments-issue-repro-220906>
- **rustfmt configuration**:
  ```toml
  format_code_in_doc_comments = true
  ```

## Description

After enabling `format_code_in_doc_comments`, rustfmt formats the following line comments to block comments – which is neither a documented nor an expected behavior.

```rust
struct TestStruct {
    position_currency: String, // Currency for position of this contract. If not null, 1 contract = 1 positionCurrency.
    pu: Option<i64>, // Previous event update sequense ("u" of previous message), -1 also means None
}
```

## Expectation

`rustfmt` should not touch this.

```rust
struct TestStruct {
    position_currency: String, // Currency for position of this contract. If not null, 1 contract = 1 positionCurrency.
    pu: Option<i64>, // Previous event update sequense ("u" of previous message), -1 also means None
}
```

## Actual Result

```rust
struct TestStruct {
    position_currency: String, /* Currency for position of this contract. If not null, 1 contract = 1 positionCurrency. */
    pu: Option<i64>, /* Previous event update sequense ("u" of previous message), -1 also means None */
}
```

## Screenshot

rustfmt with `format_code_in_doc_comments` changes the long comments to block comments

## Related

#3348

### PR DESCRIPTION
**Pull request by pan93412**
This pull request adapts the plan 2 of [@ytmimi's comment](https://github.com/rust-lang/rustfmt/issues/5533#issuecomment-1238224735), and it won't touch any comment which is not a documentation comment now.

However, it did not fixed the root cause (which is partially indicated in Plan 1): [`block_comment` is `true` if the length is too long…?](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/lists.rs#L363). As it seems not related to #5533, it may be fixed in another PR.

image

### CODE REVIEW

**Comment by pan93412** [comment]
cc @ytmimi . Thanks!

File: src/comment.rs [hunk 1]
```diff
@@ -362,24 +362,36 @@ fn identify_comment(
...
+    } else {
+        rewrite_comment_inner(
+            first_group,
+            block_style,
+            style,
+            shape,
+            config,
+            is_doc_comment_with_our_determination,
+        )?
+    };
```
Comment (ytmimi) [inline src/comment.rs:394]: I appreciate you trying to add some descriptive variable names here, but could you please try and minimize the changes that need to be made in order to fix the issue. Correct me if I'm wrong, but I think the following would have resolved the original issue.

```diff
        } else if !config.normalize_comments()
            && !config.wrap_comments()
-            && !config.format_code_in_doc_comments()
+            && (!is_doc_comment || !config.format_code_in_doc_comments())
```

File: tests/source/issue_5533.rs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// rustfmt-format_code_in_doc_comments: true
+
+struct TestStruct {
+    position_currency: String, // Currency for position of this contract. If not null, 1 contract = 1 positionCurrency.
+    pu: Option<i64>, // Previous event update sequense ("u" of previous message), -1 also means None
+}
```
Comment (ytmimi) [inline tests/source/issue_5533.rs:6]: The `source` file isn't necessary since no formatting changes are expected. Only the `target` file is needed to test that the input isn't changed.

**Review by ytmimi — COMMENTED** [review]
Thanks for your first contribution to rustfmt 🎉

See my inline comments for thoughts on how we can simplify the PR. Let me know if you have any questions.

File: src/comment.rs (hunk 1 above)
Comment (pan93412) [reply src/comment.rs]: Sure. I will just improve my own logic. However, clippy reports `nonminimal_bool` in this PR now, though it is not a fatal issue.

image

I may open another PR to improve the readability of this logic. Anyway, thanks for your valuable suggestion!

File: src/comment.rs [hunk 3]
```diff
@@ -362,12 +362,18 @@ fn identify_comment(
...
         if !config.normalize_comments() && has_bare_lines && style.is_block_comment() {
             trim_left_preserve_layout(first_group, shape.indent, config)?
         } else if !config.normalize_comments()
             && !config.wrap_comments()
-            && !config.format_code_in_doc_comments()
+            && !(
+                // `format_code_in_doc_comments` should only take effect on doc comments,
+                // so we only consider it when this comment block is a doc comment block.
+                is_doc_comment_with_our_determination && config.format_code_in_doc_comments()
+            )
```
Comment (ytmimi) [inline src/comment.rs:376]: `is_doc_comment` should always be `false` when we call `identify_comment` from `rewrite_comment`, so there's no need to also check the style.

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/comment.rs#L260-L267

```suggestion
            && !(
                // `format_code_in_doc_comments` should only take effect on doc comments,
                // so we only consider it when this comment block is a doc comment block.
                is_doc_comment && config.format_code_in_doc_comments()
            )
```

File: src/comment.rs [hunk 4]
```diff
@@ -377,7 +383,7 @@ fn identify_comment(
                 style,
                 shape,
                 config,
-                is_doc_comment || style.is_doc_comment(),
+                is_doc_comment_with_our_determination,
```
Comment (ytmimi) [inline src/comment.rs:386]: Please revert this change as it's not necessary for the PR.

**Review by ytmimi — COMMENTED** [review]
Thanks for the follow up!

I still believe `is_doc_comment` is the only boolean we need to include in our check to force code execution to move to `light_rewrite_comment`. I could be missing something, and if we need to check both `is_doc_comment` and `style.is_doc_comment` could you please explain why. The comment is helpful and I liked that you kept it 👍🏼

**Comment by pan93412** [comment]
> Thanks for the follow up!
> 
> I still believe `is_doc_comment` is the only boolean we need to include in our check to force code execution to move to `light_rewrite_comment`. I could be missing something, and if we need to check both `is_doc_comment` and `style.is_doc_comment` could you please explain why. The comment is helpful and I liked that you kept it 👍🏼

I have reverted it. However, why did L384 do such a double check?

https://github.com/rust-lang/rustfmt/blob/8919fd39d13b60a2c2e70e4f768cbd083b318de4/src/comment.rs#L384

Seems like this code has been refactored before, and I don't know why they wrote it.

**Comment by ytmimi** [comment]
> why did L384 do such a double check?

That decision pre-dates my time with the project. My best guess is that there was a time when both were necessary to be absolutely sure we were rewriting a doc comment. Now that we have `rewrite_comment` and `rewrite_doc_comment` as our top level  comment rewriting functions it's easier to differentiate what kind of comment we're rewriting.

> Seems like this code had been refactored before, and I don't know why they wrote it. Besides, even when I removed the determination, all the tests were still passed.

It's might be the case that `style.is_doc_comment()` is no longer needed, but it's not hurting us right now, so we'll keep it for now.

**Review by ytmimi — APPROVED** [review]
Again, appreciate your help on this. I think we're good to go here but will hold off on merging to give @calebcartwright a chance to take a look.

**Comment by ytmimi** [comment]
@pan93412 I noticed that this PR also resolves #5568 so I added a test case on your branch to validate that

**Comment by pan93412** [comment]
> @pan93412 I noticed that this PR also resolves #5568 so I added a test case on your branch to validate that

Thanks!

**Comment by lopopolo** [comment]
Hey folks, I think this PR has the ready to merge label and it resolves several issues with formatting comments. What else needs to be done to get this merged?

**Comment by mightyiam** [comment]
Thank you.
